### PR TITLE
Make a basic User model (including secure passwords) From Sign up

### PR DESCRIPTION
**Pull request by NguyenThu2808**

### CODE REVIEW

**Comment by NguyenThu2808** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -0,0 +1,177 @@
+@import 'bootstrap-sprockets';
+@import 'bootstrap';
+@import 'color.scss';
+
+/* mixins, variables, etc. */
+
+@mixin box_sizing {
+  -moz-box-sizing:    border-box;
```
Comment (huongnguyenmta) [inline app/assets/stylesheets/custom.scss]: thừa khoảng trắng sau dấu `:`

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+
+  def show
+  	@user = User.find(params[:id])
+  end
+
+  def new
+  	@user = User.new
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: check indent fiel này nhé em

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+class UsersController < ApplicationController
+
+  def show
+  	@user = User.find(params[:id])
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: - dùng `find_by` method
- so sánh `find_by` với `find`
- check trường hợp `nil`

File: app/helpers/users_helper.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+module UsersHelper
+  def gravatar_for(user, size: 80)
```
Comment (huongnguyenmta) [inline app/helpers/users_helper.rb]: dùng `gem config`, khai báo `80` trong settings

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  before_save { email.downcase! }
+  validates :name, presence: true, length: { maximum: 50 }
```
Comment (huongnguyenmta) [inline app/models/user.rb]: tương tự, khai báo `50` trong settings

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  VALID_EMAIL_REGEX = Settings.email_regex
+  validates :email, presence: true, length: { maximum: 255 }, format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
```
Comment (huongnguyenmta) [inline app/models/user.rb]: dòng dài quá em ơi (>80 ký tự), xuống dòng cho hợp lý nhé

File: app/views/shared/_error_messages.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t ".the_form_contains" %> <%= pluralize(@user.errors.count, t(".error")) %>.
+    </div>
+    <ul>
+    <% @user.errors.full_messages.each do |msg| %>
```
Comment (huongnguyenmta) [inline app/views/shared/_error_messages.html.erb]: indent

File: app/views/static_pages/about.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<% provide :title, t(".title") %>
+<!DOCTYPE html>
```
Comment (huongnguyenmta) [inline app/views/static_pages/about.html.erb]: từ dòng 2-7 bỏ đi được không em?

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
+<div class="row">
+	<aside class="col-md-4">
```
Comment (huongnguyenmta) [inline app/views/users/show.html.erb]: indent

**Comment by huongnguyenmta** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -1,4 +1,29 @@
 class UsersController < ApplicationController
+
+  def show
+    @user = User.find_by(id: params[:id])
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -1,4 +1,29 @@
...
+      redirect_to help_url
+    end
+  end
+
   def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -1,4 +1,29 @@
...
+    if @user.save
+      flash[:success] = t(".welcome")
+      redirect_to @user
+    else
+      render "new"
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: render :new

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -1,4 +1,29 @@
...
+  end
+
   def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      flash[:success] = t(".welcome")
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: t(".welcome") => t ".welcome"

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -1,4 +1,29 @@
...
+      redirect_to @user
+    else
+      render "new"
+    end
   end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password, :password_confirmation)
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: remove () of permit

File: app/helpers/users_helper.rb [hunk 15]
```diff
@@ -1,2 +1,8 @@
 module UsersHelper
+
+  def gravatar_for(user, size: Settings.gra_size)
```
Comment (huuphuc93) [inline app/helpers/users_helper.rb]: remove ()

File: app/helpers/users_helper.rb [hunk 16]
```diff
@@ -1,2 +1,8 @@
 module UsersHelper
+
```
Comment (huuphuc93) [inline app/helpers/users_helper.rb]: remove this empty line

File: app/helpers/users_helper.rb [hunk 17]
```diff
@@ -1,2 +1,8 @@
...
+  def gravatar_for(user, size: Settings.gra_size)
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    gravatar_url = Settings.gravatar_url
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
```
Comment (huuphuc93) [inline app/helpers/users_helper.rb]: remove ()

File: app/assets/stylesheets/custom.scss [hunk 18]
```diff
@@ -92,3 +110,68 @@ footer {
...
+      display: block;
+      margin-bottom: 3px;
+      line-height: 1;
+    }
+    h1 {
+      font-size: 1.4em;
+      text-align: left;
+      letter-spacing: -1px;
+      margin-bottom: 3px;
+      margin-top: 0px;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: You can combine line 137 and 136 into one line

File: app/models/user.rb [hunk 19]
```diff
@@ -1,8 +1,9 @@
 class User < ApplicationRecord
   before_save { email.downcase! }
```
Comment (huuphuc93) [inline app/models/user.rb]: { email.downcase! } => {email.downcase!}, same with all hash below

File: app/models/user.rb [hunk 20]
```diff
@@ -1,8 +1,9 @@
...
-  validates :name, presence: true, length: { maximum: 50 }
-  VALID_EMAIL_REGEX = Settings.valid_email_regex
-  validates :email, presence: true, length: { maximum: 255 }, format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
+  validates :name, presence: true, length: { maximum: Settings.length_name }
+  VALID_EMAIL_REGEX = Settings.email_regex
+  validates :email, presence: true, length: { maximum: Settings.length_email },
+  format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
```
Comment (huuphuc93) [inline app/models/user.rb]: indent 2 spaces

File: app/assets/stylesheets/custom.scss [hunk 21]
```diff
@@ -92,3 +110,68 @@ footer {
...
+  margin-bottom: 15px;
+  @include box_sizing;
+}
+
+input {
+  height: auto !important;
+}
+
+#error_explanation {
+  color: red;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Use color variable

File: app/assets/stylesheets/custom.scss [hunk 22]
```diff
@@ -92,3 +110,68 @@ footer {
...
+  ul {
+    color: red;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Same here

File: app/models/user.rb [hunk 23]
```diff
@@ -1,8 +1,9 @@
...
   has_secure_password
```
Comment (huuphuc93) [inline app/models/user.rb]: move this method after validates :password

File: app/views/users/new.html.erb [hunk 24]
```diff
@@ -1,3 +1,23 @@
 <% provide :title, t(".title") %>
 <h1><%= t (".signup") %></h1>
-<p><%= t (".content") %></p>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: signup_path) do |f| %>
+      <%= render "shared/error_messages" %>
+
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: remove all empty line

File: app/views/users/new.html.erb [hunk 25]
```diff
@@ -1,3 +1,23 @@
 <% provide :title, t(".title") %>
 <h1><%= t (".signup") %></h1>
-<p><%= t (".content") %></p>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: signup_path) do |f| %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: remove ()

File: app/views/users/new.html.erb [hunk 26]
```diff
@@ -1,3 +1,23 @@
 <% provide :title, t(".title") %>
 <h1><%= t (".signup") %></h1>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: t (".signup") => t ".signup"

File: app/views/users/show.html.erb [hunk 27]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (huuphuc93) [inline app/views/users/show.html.erb]: remove ()

File: app/controllers/users_controller.rb [hunk 28]
```diff
@@ -1,4 +1,29 @@
...
+      flash[:success] = t(".welcome")
+      redirect_to @user
+    else
+      render "new"
+    end
   end
+
+  private
+
+    def user_params
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: No need to add 2 space indentation after private keyword

**Comment by huuphuc93** [comment]
commented

File: app/helpers/users_helper.rb (hunk 16 above)
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Remove this empty line

File: app/helpers/users_helper.rb (hunk 15 above)
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Please remove optional ()

File: app/helpers/users_helper.rb [hunk 29]
```diff
@@ -1,2 +1,8 @@
 module UsersHelper
+
+  def gravatar_for(user, size: Settings.gra_size)
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Same here

File: app/helpers/users_helper.rb (hunk 17 above)
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Same here

File: app/models/user.rb [hunk 30]
```diff
@@ -1,8 +1,9 @@
...
   validates :password, presence: true, length: { minimum: 6 }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: For this file: 
1. Please move constant to settings.yml
2. Remove white space after the start and before the end of curly brace

File: app/models/user.rb (hunk 20 above)
Comment (mdmahmudurrahman) [inline app/models/user.rb]: As you break a line, please add 2 space indentation to it
```ruby
validates :email, presence: true, length: { maximum: Settings.length_email },
  format: { with: VALID_EMAIL_REGEX }, uniqueness: { case_sensitive: false }
```

File: app/views/shared/_error_messages.html.erb [hunk 31]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t ".the_form_contains" %> <%= pluralize(@user.errors.count, t(".error")) %>.
```
Comment (mdmahmudurrahman) [inline app/views/shared/_error_messages.html.erb]: Remove optional bracket
```ruby
<%= pluralize @user.errors.count, t(".error") %>
```

File: app/views/users/new.html.erb (hunk 25 above)
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Remove optional ()

File: app/views/users/show.html.erb (hunk 27 above)
Comment (mdmahmudurrahman) [inline app/views/users/show.html.erb]: Remove optional ()

File: config/locales/en.yml [hunk 32]
```diff
@@ -44,10 +44,21 @@ en:
...
+      createacc: "Create my account"
+      confirm: "Confirmation"
+    create:
+      welcome: "Welcome to the Sample App!"
+     
+  shared:
+    error_messages:
+      the_form_contains: "The form contains"
+      error: "error"
+
```
Comment (mdmahmudurrahman) [inline config/locales/en.yml]: Remove this empty line

File: config/routes.rb [hunk 33]
```diff
@@ -4,4 +4,6 @@
   get "/about", to: "static_pages#about"
   get "/contact", to: "static_pages#contact"
   get "/signup", to: "users#new"
+  post "/signup", to: "users#create"
+  resources :users
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: Specify the action here

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/controllers/users_controller.rb [hunk 34]
```diff
@@ -1,4 +1,28 @@
 class UsersController < ApplicationController
+
+  def show
+    @user = User.find_by id: params[:id]
+    if @user == nil
+      redirect_to help_url
+    end
+  end
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]:
```ruby
def show
  @user = User.find_by id: params[:id]
  return if @user
  flash[:warning] = "Your desired message"
  redirect_to help_path
end
```
You can also redirect to root_path

File: config/routes.rb (hunk 33 above)
Comment (mdmahmudurrahman) [reply config/routes.rb]:
```ruby
resources :users, only: i%(new create show)
```

File: app/controllers/users_controller.rb [hunk 35]
```diff
@@ -1,4 +1,28 @@
 class UsersController < ApplicationController
+
+  def show
+    @user = User.find_by id: params[:id]
+    if @user == nil
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: return if @user, add flash if can't find use

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/controllers/users_controller.rb [hunk 36]
```diff
@@ -1,4 +1,28 @@
...
+    if @user.save
+      flash[:success] = t ".welcome"
+      redirect_to @user
+    else
+      render :new
+    end
+  end
+
+  private
+  def user_params
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: add empty line after private

File: app/controllers/users_controller.rb [hunk 37]
```diff
@@ -1,4 +1,28 @@
...
+    params.require(:user).permit :name, :email, :password, :password_confirmation
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: Do not write more than 80 characters in a line

**Comment by huuphuc93** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/controllers/users_controller.rb [hunk 38]
```diff
@@ -1,4 +1,30 @@
 class UsersController < ApplicationController
+
+  def show
+    @user = User.find_by id: params[:id]
+    return if @user
+    flash[:warning] = t(".nouser")
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Please remove optional ()

File: app/views/static_pages/help.html.erb [hunk 39]
```diff
@@ -1,16 +1,9 @@
...
-    </p>
-  </body>
-</html>
+<h1><%= t ".title" %></h1>
+<p>
+  <%= t ".gethelpon" %>
+  <%= link_to t(".railshelp"), Settings.helprails %>.
+  <%= t ".torailshelp" %>
+  <em><%= link_to t(".railstutorial"), Settings.bookrails %></em>
+  <%= t ".book" %></a>.
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/help.html.erb]: Em, why you added this closing tag </a> ?

File: app/views/users/new.html.erb [hunk 40]
```diff
@@ -1,3 +1,18 @@
...
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Please remove optional ()

**Comment by mdmahmudurrahman** [comment]
commented

File: app/views/static_pages/help.html.erb (hunk 39 above)
Comment (NguyenThu2808) [reply app/views/static_pages/help.html.erb]: Because I changed "<a><em></em></a> to "link_to" tag. I forgot to delete it

**Comment by NguyenThu2808** [comment]
ready

**Comment by huuphuc93** [comment]
merged
The (pr) can be (merged) now
CC: [To:2076564] Nguyen Huy Hung C (93)
CC: [To:1546746] Md. Mahmudur Rahman (Mahmud)
CC: [To:1818111] Nguyen Thi Huong (94)

**Pull request merged**